### PR TITLE
Add PidFd::{kill, wait, try_wait}

### PR DESCRIPTION
**Pull request by the8472**
#117957 changed `Child` kill/wait/try_wait to use its pidfd instead of the pid, when one is available.
This PR extracts those implementations and makes them available on `PidFd` directly.

The `PidFd` implementations differ significantly from the corresponding `Child` methods:

* the methods can be called after the child has been reaped, which will result in an error but will be safe. This state is not observable in `Child` unless something stole the zombie child
* the `ExitStatus` is not kept, meaning that only the first time a wait succeeds it will be returned
* `wait` does not close stdin
* `wait` only requires `&self` instead of `&mut self` since there is no state to maintain and subsequent calls are safe

Tracking issue: #82971

### CODE REVIEW

File: library/std/src/os/linux/process.rs [hunk 1]
```diff
@@ -47,63 +48,98 @@ struct FileDesc;
...
+    /// [`Child::kill`]: process::Child::kill
+    pub fn kill(&self) -> Result<()> {
+        self.inner.kill()
+    }
+
+    /// Waits for the child to exit completely, returning the status that it exited with.
+    ///
+    /// Unlike [`Child::wait`] it does not ensure that the stdin handle is closed.
+    /// Additionally it will not return an `ExitStatus` if the child
+    /// has alrady been reaped. Instead an error will be returned.
```
Comment (cuviper) [inline library/std/src/os/linux/process.rs:72]:
```suggestion
    /// has already been reaped. Instead an error will be returned.
```

**Comment by cuviper** [comment]
If you call `child.pidfd().wait()` (or with `take_pidfd()`), doesn't that break `child.wait()`?

**Comment by the8472** [comment]
Good point. It shouldn't break it immediately since the pid won't be recycled until the fd is dropped (not confident, haven't tested this). But once one does drop it, yeah it'd become an issue.

But people do want to take out pidfds, e.g. for process management in tokio.
So either we have to make it `take_pidfd(self)` so it consumes the child or try_clone the fd instead so the child can hold onto its own copy.

**Comment by the8472** [comment]
I've replaced `take_pidfd` with `into_pidfd` which consumes the child.

**Comment by cuviper** [comment]
OK, `into_pidfd()` is better to make sure we avoid pid reuse on the child, but there's still an issue with the saved status. That is, `child.pidfd().wait()` cannot save the status, so it's lost from `child.wait()`. (Ditto `try_wait()`)

At the very least, this should be an unresolved question on the tracking issue. Apart from that, `r=me`.

**Comment by the8472** [comment]
Yes it'll be lost but in those cases calling `Child::wait` will return an error as `PidFd::wait` would, not  an incorrect value.
This is indeed new behavior, but any code using PidFd will also be new.

I'll note it in the tracking issue.

@bors r=cuviper rollup

**Comment by the8472** [comment]
fixed the import, docs build for x86_64-pc-windows-gnu

@bors r=cuviper

**Pull request merged**